Rewrite OSM as a package

mntzrr opened the pull request:
https://app.clickup.com/t/869410zy0

mntzrr committed:
Set up directory structure

mntzrr committed:
Manage and extend KOSM settings

mntzrr committed:
Add utility functions

mntzrr committed:
Display the placement on cart and product pages

mntzrr committed:
Update the cart amount when cart total changes

mntzrr committed:
Add support for WP widget

mntzrr committed:
Move redundant logic to Page

mntzrr committed:
Add support for shortcode

mntzrr committed:
Remove any empty elements so that the defaults can be used instead

mntzrr committed:
Replace anon functions with named

mntzrr committed:
Rename setting title and description for Client ID

mntzrr committed:
Add product variation info to debug

mntzrr committed:
Fix undefined variable warning

mntzrr committed:
Set custom library as dependent on Klarna SDK

MichaelBengtsson commented on src/KlarnaOnsiteMessaging.php:
```diff
@@ -0,0 +1,147 @@
+		$this->settings = new Settings( $settings );
+		$page           = new Product( $this->settings );
+		$cart           = new Cart( $this->settings );
+		$shortcode      = new Shortcode();
```
Make all of these class properties that are private and accessible through a method in the class. That way no one can override them, but they are still accessible. This will also let people unhook actions if they dont want to run them at the same time we do for example

MichaelBengtsson commented on src/KlarnaOnsiteMessaging.php:
```diff
@@ -0,0 +1,147 @@
+		);
+
+		if ( ! empty( $this->settings->get( 'data_client_id' ) ) ) {
+			wp_enqueue_script( 'klarna_onsite_messaging_sdk' );
```
Since this is a dependency of the other script, we don't need to enqueue it manually right? So could we possibly remove this?

MichaelBengtsson commented on src/Pages/Cart.php:
```diff
@@ -0,0 +1,77 @@
+	 *
+	 * @return void
+	 */
+	public function add_cart_total_input() {        ?>
```
Minor thing, but linebreak before the closing PHP tag

mntzrr committed:
PHP formatting

mntzrr committed:
Remove redundant script enqueue

mntzrr committed:
Encapsulate class properties

MichaelBengtsson commented:
@mntzrr You can prepare a release for the package and then merge it to master and create a release as well.

mntzrr committed:
Add notice if KOSM plugin is installed

mntzrr committed:
Unhook action hooks and dequeue KOSM plugin's scripts

mntzrr committed:
Remove unavailable themes

https://docs.klarna.com/on-site-messaging/additional-resources/placements/#attributes

mntzrr committed:
Unhook only if the KOSM plugin is active

mntzrr committed:
Deregister KOSM plugin scripts

mntzrr committed:
Check if function exists to prevent fatal error

Although is_plugin_active is a WP function it is not available until after admin_init has been triggered.

MichaelBengtsson commented on src/KlarnaOnsiteMessaging.php:
```diff
@@ -0,0 +1,240 @@
+		add_action( 'widgets_init', array( $this, 'init_widget' ) );
+
+		if ( class_exists( 'WooCommerce' ) ) {
+			add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
```
Since this now dequeues the script from the OSM plugin, should it not have a higher priority so that it always happens after the OSM plugin?

https://github.com/krokedil/klarna-onsite-messaging-for-woocommerce/blob/master/klarna-onsite-messaging-for-woocommerce.php#L50

mntzrr committed:
Enqueue scripts after KOSM plugin has been enqueued

Lower hook priority to ensure the dequeue of the KOSM plugin scripts happens AFTER they have been enqueued.

mntzrr committed:
Unhook KOSM plugin action hooks
